### PR TITLE
Bug 1648441 - Implement string list in C#

### PR DESCRIPTION
**Pull request by daoshengmu**
- [x] StringListMetricType C# binding.

- [x] Tests for StringListMetricType C# binding.

### CODE REVIEW

File: glean-core/csharp/Glean/Metrics/StringListMetricType.cs [hunk 1]
```diff
@@ -0,0 +1,172 @@
+﻿using Mozilla.Glean.FFI;
```
Comment (Dexterp37) [inline glean-core/csharp/Glean/Metrics/StringListMetricType.cs:1]:
```suggestion
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

﻿using Mozilla.Glean.FFI;
```

File: glean-core/csharp/Glean/Metrics/StringListMetricType.cs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+{
+    /// <summary>
+    /// This implements the developer facing API for recording string list metrics.
+    ///
+    /// Instances of this class type are automatically generated by the parsers at build time,
+    /// allowing developers to record values that were previously registered in the metrics.yaml file.
+    ///
+    /// The string list API only exposes the [add] and [set] methods, which takes care of validating the input
+    /// data and making sure that limits are enforced.
+    ///
```
Comment (Dexterp37) [inline glean-core/csharp/Glean/Metrics/StringListMetricType.cs:15]:
```suggestion
```

I know these are available in the Android docs, but the are outdated anyway. Let's not carry them over.

File: glean-core/csharp/Glean/Metrics/StringListMetricType.cs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                SetSync(value);
+            });
+        }
+
+        /// <summary>
+        /// Sets a string list to one or more metric stores in a synchronous way.
+        /// This is only to be used for the glean-ac to glean-core data migration.
+        /// </summary>
+        /// <param name="value">This is a user defined string list.</param>
+        public void SetSync(string[] value)
```
Comment (Dexterp37) [inline glean-core/csharp/Glean/Metrics/StringListMetricType.cs:92]:
```suggestion
        internal void SetSync(string[] value)
```

File: glean-core/csharp/Glean/Metrics/StringListMetricType.cs [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+            }
+
+            LibGleanFFI.glean_string_list_set(
+                        this.handle,
+                        value,
+                        value.Length);
+        }
+
+        /// <summary>
+        /// Tests whether a value is stored for the metric for testing purposes only.This function will
```
Comment (Dexterp37) [inline glean-core/csharp/Glean/Metrics/StringListMetricType.cs:106]:
```suggestion
        /// Tests whether a value is stored for the metric for testing purposes only. This function will
```

File: glean-core/csharp/Glean/Metrics/StringListMetricType.cs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        public bool TestHasValue(string pingName = null)
+        {
+            Dispatchers.AssertInTestingMode();
+
+            string ping = pingName ?? sendInPings[0];
+            return LibGleanFFI.glean_string_list_test_has_value(this.handle, ping) != 0;
+        }
+
+        /// <summary>
+        /// Returns the stored value for testing purposes only.This function will attempt to await the
```
Comment (Dexterp37) [inline glean-core/csharp/Glean/Metrics/StringListMetricType.cs:122]:
```suggestion
        /// Returns the stored value for testing purposes only. This function will attempt to await the
```

File: glean-core/csharp/Glean/Metrics/StringListMetricType.cs [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        /// last task(if any) writing to the the metric's storage engine before returning a value.
+        /// @throws [NullPointerException] if no value is stored
```
Comment (Dexterp37) [inline glean-core/csharp/Glean/Metrics/StringListMetricType.cs:124]:
```suggestion
```
We are properly documenting this with `<exception>` below.

File: glean-core/csharp/Glean/Metrics/StringListMetricType.cs [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+            );
+            JsonElement root = jsonPayload.RootElement;
+            int size = root.GetArrayLength();
+            string[] result = new string[size];
+ 
+            for (int i = 0; i < size; i++)
+            {
+                result[i] = root[i].ToString();
+            }
+            return result;
```
Comment (Dexterp37) [inline glean-core/csharp/Glean/Metrics/StringListMetricType.cs:152]: Did you try if `return JsoSerializer.Deserialize<string[]>(root);` works?

File: glean-core/csharp/Glean/Metrics/StringListMetricType.cs [hunk 8]
```diff
@@ -0,0 +1,172 @@
+﻿using Mozilla.Glean.FFI;
+using System;
+using System.Text.Json;
+
+namespace Mozilla.Glean.Private
```
Comment (Dexterp37) [inline glean-core/csharp/Glean/Metrics/StringListMetricType.cs:5]: Oh, almost forgot. Please kindly add an entry to [our changelog](https://searchfox.org/glean/source/CHANGELOG.md).

File: glean-core/csharp/GleanTests/Metrics/StringListMetricTypeTest.cs [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+                name: "string_list_metric",
+                sendInPings: new string[] { "store1" }
+            );
+            Assert.Throws<NullReferenceException>(() => stringListMetric.TestGetValue());
+        }
+
+        [Fact]
+        public void APISavesToSecondaryPings()
+        {
+            // Define a 'stringMetric' string metric, which will be stored in "store1"
```
Comment (travis79) [inline glean-core/csharp/GleanTests/Metrics/StringListMetricTypeTest.cs:144]:
```suggestion
            // Define a 'stringListMetric' string list metric, which will be stored in "store1"
```

File: glean-core/csharp/GleanTests/Metrics/StringListMetricTypeTest.cs [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+            var stringListMetric = new Private.StringListMetricType(
+                category: "telemetry",
+                disabled: false,
+                lifetime: Private.Lifetime.Application,
+                name: "string_list_metric",
+                sendInPings: new string[] { "store1", "store2" }
+            );
+
+            // Record two lists using add and set
```
Comment (travis79) [inline glean-core/csharp/GleanTests/Metrics/StringListMetricTypeTest.cs:153]:
```suggestion
            // Record two lists using Add() and Set()
```

File: glean-core/csharp/GleanTests/Metrics/StringListMetricTypeTest.cs [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+
+            // Check that data was properly recorded in the second ping.
+            Assert.True(stringListMetric.TestHasValue("store2"));
+            var snapshot = stringListMetric.TestGetValue("store2");
+            Assert.Equal(3, snapshot.Length);
+            Assert.Equal("value1", snapshot[0]);
+            Assert.Equal("value2", snapshot[1]);
+            Assert.Equal("value3", snapshot[2]);
+
+            // Use set() to see that the first list is replaced by the new list.
```
Comment (travis79) [inline glean-core/csharp/GleanTests/Metrics/StringListMetricTypeTest.cs:166]:
```suggestion
            // Use Set() to see that the first list is replaced by the new list.
```

File: glean-core/csharp/GleanTests/Metrics/StringListMetricTypeTest.cs [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+            Assert.Equal(3, snapshot2.Length);
+            Assert.Equal("other1", snapshot2[0]);
+            Assert.Equal("other2", snapshot2[1]);
+            Assert.Equal("other3", snapshot2[2]);
+        }
+
+        [Fact]
+        public void LongStringListsAreTruncated()
+        {
+            // Define a 'stringMetric' string metric, which will be stored in "store1"
```
Comment (travis79) [inline glean-core/csharp/GleanTests/Metrics/StringListMetricTypeTest.cs:180]:
```suggestion
            // Define a 'stringListMetric' string list metric, which will be stored in "store1"
```

File: glean-core/csharp/Glean/Metrics/StringListMetricType.cs (hunk 7 above)
Comment (daoshengmu) [reply glean-core/csharp/Glean/Metrics/StringListMetricType.cs]: Good idea! It works.

File: glean-core/csharp/GleanTests/Metrics/StringListMetricTypeTest.cs [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+                uploadEnabled: true,
+                configuration: new Configuration(),
+                dataDir: tempDataDir
+                );
+        }
+
+        [Fact]
+        public void APISavesToStorageByAddingThenSetting()
+        {
+            // Define a 'stringMetric' string metric, which will be stored in "store1"
```
[the remark this answers is not in the document]
Comment (daoshengmu) [reply glean-core/csharp/GleanTests/Metrics/StringListMetricTypeTest.cs]: I didn't notice `StringListMetricTypeTest.kt` is also wrong with this. I gave a fix to it as well.

**Comment by travis79** [comment]
The broken Swift formatting is due to an update to the linting tool and may be ignored for now.  I'll be working to fix it tomorrow.

**Pull request merged**